### PR TITLE
[#4759] feat(auth-ranger): Use Spark verify Ranger authorization Hive

### PR DESCRIPTION
**Pull request by xunliu**
### What changes were proposed in this pull request?

1. Shadow integration-test-common module.
2. Add Ranger authorization Hive IT, use Spark to verify Ranger authorization Hive.

### Why are the changes needed?

Fix: #4759

### Does this PR introduce _any_ user-facing change?

N/A

### How was this patch tested?

CI Passed.

### CODE REVIEW

File: catalogs/catalog-lakehouse-iceberg/src/test/java/org/apache/gravitino/catalog/lakehouse/iceberg/TestIcebergTable.java [hunk 1]
```diff
@@ -76,13 +76,13 @@ public class TestIcebergTable {
       NameIdentifier.of(META_LAKE_NAME, ICEBERG_CATALOG_NAME, ICEBERG_SCHEMA_NAME);
 
   @BeforeAll
-  private static void setup() {
+  public static void setup() {
```
Comment (FANNG1) [inline catalogs/catalog-lakehouse-iceberg/src/test/java/org/apache/gravitino/catalog/lakehouse/iceberg/TestIcebergTable.java:79]: any reason to change this?

File: spark-connector/spark-common/build.gradle.kts [hunk 2]
```diff
@@ -56,24 +56,12 @@ dependencies {
   testCompileOnly(libs.lombok)
 
   // use log from spark, spark3.3 use low version of log4j, to avoid java.lang.NoSuchMethodError: org.apache.logging.slf4j.Log4jLoggerFactory: method <init>()V not found
-  testImplementation(project(":api")) {
-    exclude("org.apache.logging.log4j")
```
Comment (FANNG1) [inline spark-connector/spark-common/build.gradle.kts:60]: why remove this? we should exclude log4j from Gravitino to avoid the conflict with Spark

File: integration-test-common/build.gradle.kts [hunk 3]
```diff
@@ -53,25 +57,37 @@ dependencies {
     exclude("org.elasticsearch")
     exclude("org.elasticsearch.client")
     exclude("org.elasticsearch.plugin")
+    exclude("com.amazonaws", "aws-java-sdk-bundle")
   }
-
   testImplementation(platform("org.junit:junit-bom:5.9.1"))
   testImplementation("org.junit.jupiter:junit-jupiter")
+  testImplementation(libs.bundles.jetty)
```
Comment (FANNG1) [inline integration-test-common/build.gradle.kts:64]: why adding this?

File: catalogs/catalog-lakehouse-iceberg/src/test/java/org/apache/gravitino/catalog/lakehouse/iceberg/TestIcebergTable.java (hunk 1 above)
Comment (xunliu) [reply catalogs/catalog-lakehouse-iceberg/src/test/java/org/apache/gravitino/catalog/lakehouse/iceberg/TestIcebergTable.java]: I rollback this change.

File: spark-connector/spark-common/build.gradle.kts (hunk 2 above)
Comment (xunliu) [reply spark-connector/spark-common/build.gradle.kts]: I rollback this change.

File: integration-test-common/build.gradle.kts (hunk 3 above)
Comment (xunliu) [reply integration-test-common/build.gradle.kts]: I rollback this change.

**Review by xunliu — COMMENTED** [review]
@FANNG1  Thank you for your help me to fix `log4j` issue.

Please help me review this PR, thanks.

File: integration-test-common/build.gradle.kts [hunk 4]
```diff
@@ -54,25 +62,39 @@ dependencies {
...
   }
+  testImplementation(libs.junit.jupiter.params)
+  testImplementation(libs.junit.jupiter.api)
+  testImplementation(libs.junit.jupiter.engine)
+  testImplementation(libs.bundles.jersey)
+}
 
-  testImplementation(platform("org.junit:junit-bom:5.9.1"))
-  testImplementation("org.junit.jupiter:junit-jupiter")
+val testShadowJar by tasks.registering(ShadowJar::class) {
```
Comment (FANNG1) [inline integration-test-common/build.gradle.kts:73]: after using `ShadowJar`, the integration test common jar will include almost all dependences, is it expected?

File: server-common/src/main/java/org/apache/gravitino/server/web/JettyServer.java [hunk 5]
```diff
@@ -260,6 +260,29 @@ private void initializeWebAppServletContextHandler() {
     servletContextHandler = new WebAppContext();
 
     boolean isUnitTest = System.getenv("GRAVITINO_TEST") != null;
+    if (isUnitTest) {
```
Comment (FANNG1) [inline server-common/src/main/java/org/apache/gravitino/server/web/JettyServer.java:263]: I'm curious about this change, you relocate the jetty package not exclude it, why couldn't find the jetty class?

File: server-common/src/main/java/org/apache/gravitino/server/web/JettyServer.java [hunk 6]
```diff
@@ -299,7 +322,7 @@ private void initializeWebAppServletContextHandler() {
     if (warFile.isDirectory()) {
       // Development mode, read from FS
       servletContextHandler.setResourceBase(warFile.getPath());
-      servletContextHandler.setContextPath("/");
+      servletContextHandler.addServlet(DefaultServlet.class, "/");
```
Comment (FANNG1) [inline server-common/src/main/java/org/apache/gravitino/server/web/JettyServer.java:325]: any reason changing this?

File: integration-test-common/build.gradle.kts [hunk 7]
```diff
@@ -29,9 +31,15 @@ dependencies {
   testImplementation(project(":api"))
   testImplementation(project(":clients:client-java"))
   testImplementation(project(":common"))
-  testImplementation(project(":core"))
-  testImplementation(project(":server"))
-  testImplementation(project(":server-common"))
+  testImplementation(project(":core")) {
+    exclude(group = "org.rocksdb", module = "rocksdbjni")
```
Comment (FANNG1) [inline integration-test-common/build.gradle.kts:35]: any reason exclude `rocksdb`? since this is used for test only.

File: core/src/test/java/org/apache/gravitino/connector/authorization/TestAuthorization.java [hunk 8]
```diff
@@ -73,7 +73,7 @@ public static void setUp() throws Exception {
 
     mySQLCatalog =
         new TestCatalog()
-            .withCatalogConf(ImmutableMap.of(Catalog.AUTHORIZATION_PROVIDER, "mysql"))
+            .withCatalogConf(ImmutableMap.of(Catalog.AUTHORIZATION_PROVIDER, "mysqlTest"))
```
Comment (FANNG1) [inline core/src/test/java/org/apache/gravitino/connector/authorization/TestAuthorization.java:76]: use a static variable? since it's shared by different classes.

File: integration-test-common/build.gradle.kts (hunk 4 above)
Comment (xunliu) [reply integration-test-common/build.gradle.kts:73]: The `integration-test-comman` ShadowJar main use to running `MiniGravitino` service.
We can expend a single `MiniGravitino` module in the Gravitino project in the future.
@mchades What do you think?

File: server-common/src/main/java/org/apache/gravitino/server/web/JettyServer.java (hunk 6 above)
Comment (xunliu) [reply server-common/src/main/java/org/apache/gravitino/server/web/JettyServer.java:325]: We use a `web.xml` to initial the servletContextHandler. So need change this line.

File: integration-test-common/build.gradle.kts (hunk 7 above)
Comment (xunliu) [reply integration-test-common/build.gradle.kts:35]: The `rocksdb-xxx.jar` have 47MB size, let Shadow jar package very big, and currently we doesn't support rocksdb storage. So I exclude it.

File: core/src/test/java/org/apache/gravitino/connector/authorization/TestAuthorization.java (hunk 8 above)
Comment (xunliu) [reply core/src/test/java/org/apache/gravitino/connector/authorization/TestAuthorization.java:76]: Not other place needs to use this static variable.

File: authorizations/authorization-ranger/src/test/java/org/apache/gravitino/authorization/ranger/integration/test/RangerHiveE2EIT.java [hunk 9]
```diff
@@ -77,63 +78,207 @@ public class RangerHiveE2EIT extends AbstractIT {
...
+              "__REPLACE__RANGER_HIVE_REPO_NAME", RangerITEnv.RANGER_HIVE_REPO_NAME);
+    } finally {
+      inputStream.close();
+    }
+
+    FileOutputStream outputStream = new FileOutputStream(xmlPath);
+    try {
+      IOUtils.write(templateContext, outputStream, StandardCharsets.UTF_8);
+    } finally {
+      outputStream.close();
```
Comment (yuqi1129) [inline authorizations/authorization-ranger/src/test/java/org/apache/gravitino/authorization/ranger/integration/test/RangerHiveE2EIT.java:192]:
```suggestion
      String templateContext = FileUtils.readFileToString(new File(templatePath), StandardCharsets.UTF_8);
      templateContext = templateContext.replace("__REPLACE__RANGER_ADMIN_URL", RANGER_ADMIN_URL)
          .replace("__REPLACE__RANGER_HIVE_REPO_NAME", RangerITEnv.RANGER_HIVE_REPO_NAME);
      FileUtils.writeStringToFile(new File(xmlPath), templateContext, StandardCharsets.UTF_8);
```

File: integration-test-common/build.gradle.kts [hunk 10]
```diff
@@ -54,25 +61,41 @@ dependencies {
...
     exclude("org.elasticsearch.client")
     exclude("org.elasticsearch.plugin")
+    exclude("com.amazonaws", "aws-java-sdk-bundle")
   }
+  testImplementation(libs.apiguardian.api)
+  testImplementation(libs.bundles.metrics)
+  testImplementation(libs.junit.jupiter.params)
+  testImplementation(libs.junit.jupiter.api)
+  testImplementation(libs.junit.jupiter.engine)
+  testImplementation(libs.bundles.jersey)
```
Comment (yuqi1129) [inline integration-test-common/build.gradle.kts:71]: Please keep them in alphabetical order.

File: core/src/main/java/org/apache/gravitino/storage/relational/database/H2Database.java [hunk 11]
```diff
@@ -54,6 +54,12 @@ public String startH2Database(Config config) {
 
     String connectionUrl = constructH2URI(originalJDBCUrl, storagePath);
 
+    try {
+      Class.forName("org.h2.Driver");
```
Comment (yuqi1129) [inline core/src/main/java/org/apache/gravitino/storage/relational/database/H2Database.java:58]: Why do we need to explicitly load the driver? I remember that the driver-jar is already in the class path.

File: server-common/src/main/java/org/apache/gravitino/server/web/JettyServer.java (hunk 5 above)
Comment (FANNG1) [reply server-common/src/main/java/org/apache/gravitino/server/web/JettyServer.java:263]: After remove `server-common` test dependences in other modules, this seems unnecessary anymore?

File: core/src/main/java/org/apache/gravitino/storage/relational/database/H2Database.java (hunk 11 above)
Comment (xunliu) [reply core/src/main/java/org/apache/gravitino/storage/relational/database/H2Database.java:58]: Class.forName(`org.h2.Driver`) is called to load and register the JDBC driver.  
Before JDBC 4.0, Java applications needed to explicitly load the JDBC driver to be used, which was done by calling the `Class.forName()` method and providing the full class name of the driver. 
This call causes the driver class to be loaded into the JVM, and then the driver class's static initializer registers the driver itself with the `java.sql.DriverManager` class so that the DriverManager can use it to establish a database connection.  Without this call, when you try to establish a database connection via `DriverManager.getConnection()`, the DriverManager may not find a suitable driver to connect to the database and therefore throws an exception.

File: integration-test-common/build.gradle.kts (hunk 10 above)
Comment (xunliu) [reply integration-test-common/build.gradle.kts:70]: DONE

**Review by xunliu — COMMENTED** [review]
@FANNG1 @yuqi1129 
Please help me review again, thanks.

File: gradle/libs.versions.toml [hunk 12]
```diff
@@ -217,7 +218,7 @@ datanucleus-core = { group = "org.datanucleus", name = "datanucleus-core", versi
 datanucleus-api-jdo = { group = "org.datanucleus", name = "datanucleus-api-jdo", version.ref = "datanucleus-api-jdo" }
 datanucleus-rdbms = { group = "org.datanucleus", name = "datanucleus-rdbms", version.ref = "datanucleus-rdbms" }
 datanucleus-jdo = { group = "org.datanucleus", name = "javax.jdo", version.ref = "datanucleus-jdo" }
-
+apiguardian-api = { group = "org.apiguardian", name = "apiguardian-api", version.ref = "apiguardian-api" }
```
Comment (mchades) [inline gradle/libs.versions.toml:221]: Need to add to `LICENSE.bin`?

File: core/src/main/java/org/apache/gravitino/storage/relational/database/H2Database.java (hunk 11 above)
Comment (mchades) [reply core/src/main/java/org/apache/gravitino/storage/relational/database/H2Database.java:58]: I am confused about why it worked before without needing an explicit call?

File: integration-test-common/src/test/java/org/apache/gravitino/integration/test/container/RangerContainer.java [hunk 13]
```diff
@@ -38,7 +38,7 @@ public class RangerContainer extends BaseContainer {
   public static final String DEFAULT_IMAGE = System.getenv("GRAVITINO_CI_RANGER_DOCKER_IMAGE");
   public static final String HOST_NAME = "gravitino-ci-ranger";
   public static final int RANGER_SERVER_PORT = 6080;
-  public RangerClientExtension rangerClient;
+  public RangerClient rangerClient;
```
Comment (mchades) [inline integration-test-common/src/test/java/org/apache/gravitino/integration/test/container/RangerContainer.java:41]: I saw that this variable was originally used `RangerClient`, then you changed it to `RangerClientExtension`, why did you change it back to `RangerClient`? Did I miss something?

File: integration-test-common/build.gradle.kts (hunk 10 above)
Comment (mchades) [reply integration-test-common/build.gradle.kts:70]: not done

File: integration-test-common/build.gradle.kts (hunk 4 above)
Comment (mchades) [reply integration-test-common/build.gradle.kts:73]: Yeah, just like mini HMS. It can be imported and setup directly by other programs without introducing additional dependencies.

File: server-common/src/main/java/org/apache/gravitino/server/web/JettyServer.java (hunk 5 above)
Comment (mchades) [reply server-common/src/main/java/org/apache/gravitino/server/web/JettyServer.java:263]: @xunliu need answer.

Comment (mchades) [reply server-common/src/main/java/org/apache/gravitino/server/web/JettyServer.java:263]: Why does only UT need this?

File: .github/workflows/python-integration-test.yml [hunk 14]
```diff
@@ -48,7 +48,7 @@ jobs:
     needs: changes
     if: needs.changes.outputs.source_changes == 'true'
     runs-on: ubuntu-latest
-    timeout-minutes: 30
+    timeout-minutes: 45
```
Comment (FANNG1) [inline .github/workflows/python-integration-test.yml:51]: use a seperate PR to change this?

File: authorizations/authorization-ranger/build.gradle.kts [hunk 15]
```diff
@@ -57,36 +61,26 @@ dependencies {
     exclude("net.java.dev.jna")
     exclude("javax.ws.rs")
     exclude("org.eclipse.jetty")
+    exclude("com.amazonaws", "aws-java-sdk-bundle")
```
Comment (FANNG1) [inline authorizations/authorization-ranger/build.gradle.kts:64]: Could you provide a sperate PR?

File: authorizations/authorization-ranger/src/test/java/org/apache/gravitino/authorization/ranger/integration/test/RangerHiveE2EIT.java [hunk 16]
```diff
@@ -77,63 +76,195 @@ public class RangerHiveE2EIT extends AbstractIT {
...
-            containerSuite.getHiveContainer().getContainerIpAddress(),
+            containerSuite.getHiveRangerContainer().getContainerIpAddress(),
             HiveContainer.HIVE_METASTORE_PORT);
 
+    generateRangerSparkSecurityXML();
+
+    sparkSession =
+        SparkSession.builder()
+            .master("local[1]")
+            .appName("Hive Catalog integration test")
```
Comment (FANNG1) [inline authorizations/authorization-ranger/src/test/java/org/apache/gravitino/authorization/ranger/integration/test/RangerHiveE2EIT.java:131]: better add ranger to app name

File: core/src/main/java/org/apache/gravitino/storage/relational/database/H2Database.java (hunk 11 above)
Comment (FANNG1) [reply core/src/main/java/org/apache/gravitino/storage/relational/database/H2Database.java:58]: +1, this may be caused by shade of integration test common

**Comment by FANNG1** [comment]
I'm afraid the current shade implement seems to lead some unexpected behavior,  makes the test class more complicated.

**Comment by FANNG1** [comment]
BTW, I prefer to drop the support for MiniGravitino to avoid to struggle for the class conflict related problems. @mchades @yuqi1129 @xunliu WDYT?

**Comment by mchades** [comment]
> BTW, I prefer to drop the support for MiniGravitino to avoid to struggle for the class conflict related problems. @mchades @yuqi1129 @xunliu WDYT?

Do you mean to remove the embedded tests?

**Comment by FANNG1** [comment]
> Do you mean to remove the embedded tests?

Yes, MiniGravitino provides the facilities to debug easily, but leading complicated class conflict problems, it takes much more time to debug to me.

File: integration-test-common/src/test/java/org/apache/gravitino/integration/test/container/RangerContainer.java (hunk 13 above)
Comment (xunliu) [reply integration-test-common/src/test/java/org/apache/gravitino/integration/test/container/RangerContainer.java:41]: Because `RangerClientExtension` in the `authorization-ranger` module.
I want the `integration-test-common` not to depend `authorization-ranger` module.

**Comment by xunliu** [comment]
> > Do you mean to remove the embedded tests?
> 
> Yes, MiniGravitino provides the facilities to debug easily, but leading complicated class conflict problems, it takes much more time to debug to me.

I think MiniGravitino is still a necessity.
1. Other systems will also require the use of MiniGravitino for integration testing. Like MiniHiveMetastore.
2. Without miniGravitino, community developers would have to use remote debugging mode, which is too difficult.

I think the best way to do this is that we should create a separate MiniGravitino module.

**Comment by xunliu** [comment]
@mchades 
> I am confused about why it worked before without needing an explicit call?

Because the module after the shadow, there is no longer this file structure `META-INF/services/java.sql.Driver` , so you can not find the h2database through the SPI, you must show the call `Class.forName("org.h2.Driver");` function

image

**Comment by yuqi1129** [comment]
> @mchades
> 
> > I am confused about why it worked before without needing an explicit call?
> 
> Because the module after the shadow, there is no longer this file structure `META-INF/services/java.sql.Driver` , so you can not find the h2database through the SPI, you must show the call `Class.forName("org.h2.Driver");` function
> 
> image

@xunliu 
The file exists, but the content in it is unexpected, please see https://github.com/apache/gravitino/pull/4948#discussion_r1772694796

**Comment by xunliu** [comment]
> @xunliu need answer.

Because Jetty server default load https://github.com/jetty/jetty.project/blob/jetty-12.0.x/jetty-core/jetty-deploy/src/test/resources/etc/webdefault.xml#L37, But we rename `org.eclipse.jetty` to `org.apache.gravitino.it.shaded.org.eclipse.jetty`, So we load `server-common/src/test/resources/web.xml` in the test mode.

Don't worry, the Jetty server loads `web.xml` in the web module in the produce mode, There's no such thing.

File: core/src/main/java/org/apache/gravitino/storage/relational/database/H2Database.java (hunk 11 above)
Comment (xunliu) [reply core/src/main/java/org/apache/gravitino/storage/relational/database/H2Database.java:58]: I used `mergeServiceFiles()` to fix this problem.

**Comment by xunliu** [comment]
@FANNG1 @mchades @yuqi1129 I improved this PR based on your suggestion, Please help me review it again, Thanks.

**Comment by xunliu** [comment]
@FANNG1 @mchades @yuqi1129 I refactored this PR, Please help me review it again, thanks.

File: authorizations/authorization-ranger/src/test/java/org/apache/gravitino/authorization/ranger/integration/test/RangerITEnv.java [hunk 17]
```diff
@@ -106,11 +108,30 @@ public static void cleanup() {
     }
   }
 
+  static void startHiveRangerContainer() {
+    containerSuite.startHiveRangerContainer(
+        new HashMap<>(
+            ImmutableMap.of(
```
Comment (yuqi1129) [inline authorizations/authorization-ranger/src/test/java/org/apache/gravitino/authorization/ranger/integration/test/RangerITEnv.java:114]: Why do we need a nested Map?

**Review by yuqi1129 — COMMENTED** [review]
LGTM except a minor one.

**Review by FANNG1 — APPROVED** [review]

File: authorizations/authorization-ranger/src/test/java/org/apache/gravitino/authorization/ranger/integration/test/RangerITEnv.java (hunk 17 above)
Comment (xunliu) [reply authorizations/authorization-ranger/src/test/java/org/apache/gravitino/authorization/ranger/integration/test/RangerITEnv.java:114]: Yes, you are right, I already fixed it in the next PR.
